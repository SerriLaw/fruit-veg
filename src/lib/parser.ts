import { groupBy, map, uniq } from 'lodash';

import { Item } from '../types/item';
import data from '../data/fruit-and-veg.json';

export default function parseData() {
  return group(
    data.map(x => ({
      type: x.type,
      name: x.name,
      months: [x.month],
      seasons: [x.season]
    }))
  );
}

function group(data: any[]): Item[] {
  return map(groupBy(data, z => z.name), group => {
    return group.reduce((i, y) => ({
      name: i.name,
      type: i.type,
      months: [...i.months, ...y.months],
      seasons: uniq([...i.seasons, ...y.seasons])
    }));
  });
}
