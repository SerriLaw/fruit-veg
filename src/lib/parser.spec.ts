import { transformData } from './parser';

const data = [
  {
    name: 'Apples',
    type: 'fruit',
    season: 'autumn',
    month: 'march'
  },
  {
    name: 'Bananas',
    type: 'fruit',
    season: 'autumn',
    month: 'april'
  },
  {
    name: 'Potatoes',
    type: 'vegetable',
    season: 'spring',
    month: 'september'
  },
  {
    name: 'Bananas',
    type: 'fruit',
    season: 'autumn',
    month: 'march'
  },
  {
    name: 'Potatoes',
    type: 'vegetable',
    season: 'summer',
    month: 'december'
  }
];

describe('Parser', () => {
  it('should transform raw data into Item arrays', () => {
    expect(transformData(data)).toMatchSnapshot();
  });
});
