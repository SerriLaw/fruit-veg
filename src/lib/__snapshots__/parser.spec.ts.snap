// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser should transform raw data into Item arrays 1`] = `
Array [
  Object {
    "months": Array [
      "march",
    ],
    "name": "Apples",
    "seasons": Array [
      "autumn",
    ],
    "type": "fruit",
  },
  Object {
    "months": Array [
      "april",
      "march",
    ],
    "name": "Bananas",
    "seasons": Array [
      "autumn",
    ],
    "type": "fruit",
  },
  Object {
    "months": Array [
      "september",
      "december",
    ],
    "name": "Potatoes",
    "seasons": Array [
      "spring",
      "summer",
    ],
    "type": "vegetable",
  },
]
`;
